### ISSUE TITLE
`length-zero-no-unit` should warn for percentages

### ISSUE THREAD

**Issue by XhmikosR** [stylelint/stylelint#2945]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Bug.

> Which rule, if any, is this issue related to?

`length-zero-no-unit`

> What CSS is needed to reproduce this issue?

```css
.card-title {
  margin-bottom: 0%;
}
```

> What stylelint configuration is needed to reproduce this issue?

```json
{
  "rules": {
    "length-zero-no-unit": true
  }
}
```

> Which version of stylelint are you using?

`8.1.1`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

CLI with `stylelint --config myconfig *.css`

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Nope

> What did you expect to happen?

A warning should be thrown about using a unit with zero (`0%`).

> What actually happened (e.g. what warnings or errors you are getting)?

No warning.

-----

There are some cases a percentage needs to be specified for 0, like for [flex bug `#4`](https://github.com/philipwalton/flexbugs#4-flex-shorthand-declarations-with-unitless-flex-basis-values-are-ignored). There might be more cases which I don't recall atm.

**Comment by jeddy3** [comment]
@XhmikosR Thanks for writing up the issue and for using the template.

> Bug.

I think the `length-zero-no-unit` rule is behaving as expected here because [_percentages_](https://www.w3.org/TR/css-values-3/#percentages) aren't [_lengths_](https://www.w3.org/TR/css-values-3/#lengths). This rule [specifically](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/about-rules.md#about-rule-names) disallows units for zero _lengths_. 

I believe what you're after is a new rule that enforces using a _length_ of zero instead of a _percentage_ of zero _when possible_ e.g. `margin-left` [can have](https://developer.mozilla.org/en-US/docs/Web/CSS/margin-left#Values) a value that is a _length_ or a _percentage_ (or the `auto` keyword), so the new rule would enforce a _length_ of zero here. However, I believe there are a few places in the spec where this isn't the case e.g. [the parameters of the `hsl()` color function](https://drafts.csswg.org/css-color-4/#funcdef-hsl) can only be _percentages_, so a parameter of `0` would not be valid here. This makes the rule somewhat tricky to implement, and I assume this is why it hasn't (unlike the `length-zero-no-unit` rule) been implemented in stylelint yet. Does that make sense?

It could be that someone knocks up a rough version of the rule in an attempt to uncover any potential issues with it. Based on that we can decide whether it's [best suited](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#criteria-for-inclusion) as a built-in rule or a plugin. It could be a good first plugin/rule for someone as it will likely require trawling through the specs to learn about the where _lengths_ and _percentages_ are used.

> There are some cases a percentage needs to be specified for 0, like for flex bug #4. There might be more cases which I don't recall atm.

See https://github.com/stylelint/stylelint/issues/2522 for our recommended approach for dealing with things like `flex: 1 0 0px`. Such an approach would also be recommended way for the new rule mentioned above.

**Comment by CAYdenberg** [comment]
I'm in agreement that formally percentages are not lengths, but I can't really see why the distinction matters from the users' perspective. (vh and vw are also relative to other lengths, and arguably em and rem are too, so really the only formal distance unit in CSS is px). I don't know why a stylelint user would want `0%` to be ok within the context of a "distance" rule but not `0px`.

I agree that implentation could be extraordinarily complex (considering rules like "box-shadow" that have mixes of colors and distances, which can be in different positions) ... BUT I think that in an ideal world were this to be implemented, it would not be a new rule but merely a refinement of this one?

**Comment by jeddy3** [comment]
> but I can't really see why the distinction matters from the users' perspective

Yes, I agree that, on the surface, the distinction doesn't matter to the user. However, the approach to how rules are built and named is not only about usability. It was designed to find a balance between usability and the sustainability of the project. (Searching for "sustainable" in this repo should dig up some of those historical discussions.) Paradoxically, I believe pursuing this balance _improves usability_ as users can be confident that all the built-in rules _work completely_ (for standard syntax) i.e. they faultless do exactly what their names say they will do.  I also assume this is why there are a manageable number of bugs against the built-in rules. Taking the `length-zero-no-unit` rule as an example, it was scoped to _lengths_ because that implementation was finishable and achievable within a reasonable amount of time.

> BUT I think that in an ideal world were this to be implemented,

I think you hit the nail on the head here. It's _not_ an ideal world, so we came up with:

1. The list of [inclusion criteria](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/rules.md#criteria-for-inclusion), including a "Has a clear and unambiguous finished state".
2. Laser-focused rules (e.g. we have separate `*-no-vendor-prefixes` and `*-no-duplicates`), rather than singular rules (e.g. `no-vendor-prefixes` and `no-duplicates`) with options, so that rules can more easily adhere to this criteria.

Having said all that, if you believe a singular rule for percentages and lengths is beneficial then a plugin could be created for this. It could even [build on](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/plugins.md#stylelintutilscheckagainstrule) of the built-in `length-zero-no-unit`, and add the logic to determine where percentages can also be zero lengths.

However, my suggestion is a separate (and experimental) `percentage-no-zero` plugin that contains just this logic (and relevant tests). The plugin can always be moved into core once the inclusion criteria are met.

**Comment by CAYdenberg** [comment]
Alright thanks for the gentle reminder that there's a history to all of this :+1: 

With those community values in mind I'm in agreement with your approach. I suppose if that's the case we close this?

**Comment by jeddy3** [comment]
> I suppose if that's the case we close this?

Yep, I think we can close this issue as a plugin.

If either yourself, @XhmikosR or anyone else is interested in building this plugin, then there's [a section](https://github.com/stylelint/stylelint/blob/master/docs/developer-guide/plugins.md) in the Developer Guide with instructions. It could be a fascinating plugin to attempt to build. It'll be interesting to see if it's achievable and if the complexities around `hsl` colours and `box-shadow` values can be tamed.

- Name: `stylelint-percentage-no-zero`
- Description: Disallow zero percentages where a zero length can be used.
- Primary option: `true`
- Secondary options: none
- Message: "Unexpected zero percentage"

And as I said, this plugin could be moved into core once the inclusion criteria are met.

**Comment by jbruni** [comment]
Guys... I'm here to tell you that I have a page where an element with style `height: 0%;` renders different from having style `height: 0;` - using latest Chrome (Version 66.0.3359.181 - Official Build - 64-bit).

I can provide more details if necessary.

**Comment by yisibl** [comment]
> I'm in agreement that formally percentages are not lengths, but I can't really see why the distinction matters from the users' perspective. (vh and vw are also relative to other lengths, and arguably em and rem are too, so really the only formal distance unit in CSS is px). I don't know why a stylelint user would want `0%` to be ok within the context of a "distance" rule but not `0px`.

@CAYdenberg You can see the difference here: https://codepen.io/judowalker/pen/xrMxWj
